File: src/psilox/core/Psilox.java
package psilox.core;

import psilox.audio.Audio;
import psilox.node.Node;
import psilox.utility.Time;

public class Psilox {
	
	private static final long DEFAULT_INTERVAL = Time.SECOND / 60;
	
	private static Window window;
	
	private static boolean running;
	
	private static long tickInterval = DEFAULT_INTERVAL;
	private static long frameInterval = DEFAULT_INTERVAL;
	public static float dt, du, df;
	public static long ticks;
	
	private static Thread psiloxThread;
	
	public static final Node root = new Node("root");
	private static Node mainNode;
	private static Node nextScene;
	
	public static void start(Window window, Node mainNode) {
		if(running) return;
		Psilox.window = window;
		Psilox.mainNode = mainNode;
		initThread();
	}
	
	public static void quit() {
		running = false;
	}
	
	private static void update() {
		window.pollEvents();
		root.updateChildren();
		Node.applyChanges();
	}
	
	private static void render() {
		window.clear();
		
		window.renderTree(root);
		
		window.logLastError();
		window.swapBuffers();
	}
	
	private static void loop() {
		long lastTick = Time.now() - tickInterval;
		long lastFrame = Time.now() - frameInterval;
		
		window.initialize();
		Audio.init();
		root.addChild(mainNode);
		
		while(running) {
			if(nextScene != null) {
				root.swapChild(nextScene, 0);
				nextScene = null;
			}
			if(Time.since(lastTick) >= tickInterval) {
				dt = Time.since(lastTick) / (float) Time.SECOND;
				lastTick = Time.now();
				update();
				ticks++;
				du = (Time.now() - lastTick) / (float) Time.SECOND;
			}
			
			if(Time.since(lastFrame) >= frameInterval) {
				lastFrame = Time.now();
				render();
				df = (Time.now() - lastFrame) / (float) Time.SECOND;
			}
			
			running &= !window.shouldClose();
		}
		
		Audio.shutdown();
		window.terminate();
	}
	
	public static void changeScene(Node nextScene) {
		Psilox.nextScene = nextScene;
	}
	
	private static void initThread() {
		running = true;
		if(System.getProperty("os.name").contains("Mac")) {
			System.setProperty("java.awt.headless", "true");
			loop();
		} else {
			(psiloxThread = new Thread(Psilox::loop, "psilox")).start();
		}
	}
	
}

File: src/psilox/core/Psilox.kt
package psilox.core

import psilox.graphics.RED
import psilox.resource.ResourceLibrary
import psilox.utility.Time
import psilox.utility.Time.elapsed
import psilox.utility.Time.now
import psilox.utility.Time.per
import psilox.utility.Time.sinceF

object Psilox {

    lateinit var window: Window
    var running = false

    private var tickInterval = 60 per Time.SECOND
    private var frameInterval = 60 per Time.SECOND

    var dt = 0F; private set
    var ticks = 0L; private set

    fun start(window: Window) {
        if (running) return
        running = true
        this.window = window
        System.setProperty("java.awt.headless", "true")

        loop()
    }

    fun quit() {
        running = false
    }

    private fun update() {
        window.pollEvents()
        //root.updateChildren()
        //Node.applyChanges()
    }

    private fun render() {
        window.clear()
        //window.renderTree(root);
        window.logLastError();
        window.swapBuffers();
    }

    private fun loop() {
        var lastTick = 0L
        var lastFrame = 0L

        window.initialize()

        while(running) {
            if(lastTick elapsed tickInterval) {
                dt = sinceF(lastTick, Time.SECOND)
                lastTick = now
                update()
                ticks++
            }

            if(lastFrame elapsed frameInterval) {
                lastFrame = now
                render()
            }

            ResourceLibrary.clearDereferenced()
            running = running and !window.shouldClose()
        }

        ResourceLibrary.clearAll()
        window.terminate()
    }

}

fun main(args: Array<String>) {
    Psilox.start(Window(clear = RED))
}